### ISSUE TITLE
Add an assist to generate documentation templates

### ISSUE THREAD

**Issue by numero-744** [rust-analyzer/rust-analyzer#10878]
This assist would allow to generate a documentation template containing sections according to https://rust-lang.github.io/api-guidelines/documentation.html above an undocumented public function.

The goal is not to generate the whole documentation for a function, it is just to guide the author documenting the function.

PR: #10951 

## Example

```rust
fn my_function(a: i32, b: SomeRandomType) -> Result<OtherRandomType> { ... }
```

Would generate

```rust
/// .
///
/// # Examples
///
/// ```rust
/// use my_crate::my_function;
/// assert_eq!(my_function(a, b), );
/// ```
///
/// # Errors
///
/// This function will return an error if .
```

A few specifications are written below. They are not exhaustive and need discussion.

- [x] This new assist is unit tested
- [x] This new assist is not applicable in a trait implementation as trait-associated functions should be documented in the trait, not in its implementations.
- [x] This new assists writes the start of a description for `new() -> Self`

## C-EXAMPLE

- [x] It would always contain an `# Example` section, adding a pre-filled snippet of code calling the function.
- [x] The example would start with `use my_crate::...;`
- [x] Arguments would be declared first if they are `self` or `&mut`, without initialization to let the author give a relevant value to the argument. Mutability would be managed (for instance a `fn function(&mut self)` would  generate something like `let mut something;` where "something" can be the sneaky_cased name of the Type implemented, see the item below. For arguments other than `self`, the variable name would be the same as the argument name.
- [x] The function call would be a method call if the first argument is a `self`.
- [x] The function call would be `MyTrait::my_func()` if it is a trait-associated and not a method call.
- [x] The function call would be followed by an `assert_eq!` for its return value, to check the returned value.
- [x] The function call would be followed by an `assert_eq!` for each of its `&mut` argument to check the new value of the argument.
- [x] If the function returns something, the function is the RHS of a `let result =`.
- [x] If the function has less than 3 arguments, the function is called directly as the first argument of `assert_eq!`, without a `let`.
- [x] For an unsafe function, the function call would be put into an `unsafe` block.
- [x] If the function is guessed to panic in some situations, an example which should panic is added.

## C-FAILURE

- [x] If the return type of the function _contains_ `Result` (simplified version to detect a result), an `# Errors` section would be added, with a text template "This function will return an error if"
- [x] If the function is `unsafe`, a `# Safety` section would be added
- [x] [Discussion needed] If the function contains an `.unwrap()`, `.expect()`, `panic!()` or `assert!()`, a `# Panics` section would be added. (It would be nice to guess if the function may panic because of inappropriate input or state.)

## Discussion

- [x] In the generated example, maybe the arguments could be defined for instance if the type is `Default` or has a `::new()` function.
- [x] In the generated example, when the function call is moved out of the `assert_eq!`, the name of the returned value could be discussed.
- [x] When to add a `# Panics` section?

Comments appreciated :smiley:

**Comment by numero-744** [comment]
It seems to happen in [`rust-analyzer/crates/ide_assists/src/lib.rs`](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide_assists/src/lib.rs) where a [handler](https://github.com/rust-analyzer/rust-analyzer/tree/master/crates/ide_assists/src/handlers) may be added, for instance named `generate_doc_template`. This handler could be modified later to be able to generate documentation patterns for other items if relevant.

May I start working on this feature in a fork?

**Comment by Veykril** [comment]
Note there is already an issue regarding this with a few bits of discussion as well here -> https://github.com/rust-analyzer/rust-analyzer/issues/4767 
and another issue(with less thought behind it) similar to this https://github.com/rust-analyzer/rust-analyzer/issues/1915

**Comment by Veykril** [comment]
Adding such an assist seems fine to me for the time being,
but as matklad stated in #4767 which I agree with, ideally this should probably be a completion thing(though we are nowhere close to that), in which case the work for the assist should be easily transferable I imagine.

So yes feel free to have a stab at this 👍

**Comment by numero-744** [comment]
Thanks for the clearance and for the links to other issues!
https://github.com/rust-analyzer/rust-analyzer/issues/4767 points to useful links, that I will read carefully.

I think I will start with an assist, which seems to be simpler to implement at first sight (first contribution to rust-analyser) :sweat_smile:

**Comment by numero-744** [comment]
The implementation is done, I will write the unit tests and debug if needed before submitting a PR.

I think the PR will fix this issue (#10878) and #1915.

About #4767, the PR may fix it. It points to [RFC 1574](https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md) which the suggested assist seems to follow, except for sections "Aborts" and "Undefined Behavior" which look like "Panics" and "Safety" (I may not understand the goal of these two sections) and therefore are not generated. It also points to [this article](https://deterministic.space/machine-readable-inline-markdown-code-cocumentation.html) which is in favor of Doxygen-like comments and is not approved by an RFC afaik, so I have chosen to not follow its guidelines.

For the remaining discussions of this issue (#10878), I would say:

- The RHS of argument definitions should not be guessed even if they have a `new()` or if they are `Default` because the generated RHS would not be relevant in most cases.
- The returned variable is currently named `result`, which does not seem that confusing even if it is not a `Result<_, _>` imho.
- For the case when there is already some documentation for the function, what do you think that should be done?
